lxd/identities.go

```diff
@@ -1201,46 +1201,154 @@ func updateIdentity(authenticationMethod string) func(d *Daemon, r *http.Request
-			return response.NotImplemented(fmt.Errorf("Identities of type %q cannot be modified via this API", id.Type))
+		if id.Type != api.IdentityTypeCertificateClient && identityPut.TLSCertificate != "" {
+			return response.BadRequest(fmt.Errorf("Cannot update certificate for identities of type %q", id.Type))
 		}
 
-		canViewGroup, err := s.Authorizer.GetPermissionChecker(r.Context(), auth.EntitlementCanView, entity.TypeAuthGroup)
+		err = s.Authorizer.CheckPermission(r.Context(), entity.IdentityURL(authenticationMethod, id.Identifier), auth.EntitlementCanEdit)
+		if err == nil {
+			return updateIdentityPrivileged(s, r, *id, identityPut)
+		} else if !auth.IsDeniedError(err) {
+			return response.SmartError(err)
+		}
+
+		// Only identities of type api.IdentityTypeCertificateClient may update their own certificate
+		if id.Type != api.IdentityTypeCertificateClient {
+			return response.Forbidden(nil)
+		}
+
+		username, err := auth.GetUsernameFromCtx(r.Context())
 		if err != nil {
 			return response.SmartError(err)
 		}
 
-		err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
-			apiIdentity, err := id.ToAPI(ctx, tx.Tx(), canViewGroup)
-			if err != nil {
-				return err
-			}
+		// Identities may only update their own certificate
+		if username != id.Identifier {
+			return response.Forbidden(nil)
+		}
 
-			err = util.EtagCheck(r, apiIdentity)
-			if err != nil {
-				return err
-			}
+		return updateIdentityUnprivileged(s, r, *id, identityPut)
+	}
+}
 
-			err = dbCluster.SetIdentityAuthGroups(ctx, tx.Tx(), id.ID, identityPut.Groups)
-			if err != nil {
-				return err
-			}
+// updateIdentityUnprivileged is only invoked when an identity of type api.IdentityTypeClientCertificate updates their
+// own identity and does not have permission to change their own groups.
+func updateIdentityUnprivileged(s *state.State, r *http.Request, id dbCluster.Identity, identityPut api.IdentityPut) response.Response {
+	// Validate the given certificate
+	fingerprint, metadata, err := validateIdentityCert(s.Endpoints.NetworkCert(), identityPut.TLSCertificate)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	canViewGroup, err := s.Authorizer.GetPermissionChecker(r.Context(), auth.EntitlementCanView, entity.TypeAuthGroup)
+	if err != nil {
+		return response.SmartError(err)
+	}
 
+	err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
+		apiIdentity, err := id.ToAPI(ctx, tx.Tx(), canViewGroup)
+		if err != nil {
+			return err
+		}
+
+		err = util.EtagCheck(r, apiIdentity)
+		if err != nil {
+			return err
+		}
+
+		// Return an error if the caller tries to update their own groups.
+		if !slices.Equal(identityPut.Groups, apiIdentity.Groups) {
+			return api.NewStatusError(http.StatusForbidden, "Only the certificate may be changed")
+		}
+
+		// If the fingerprint is identical there is nothing to do (this is checked here so that we still return an error
```

Comment:
this comment didnt make sense to me.

Reply:
We need to start a transaction to convert the `cluster.Identity` into an `api.Identity` and get it's list of groups. When a user without `can_edit` on the identity updates their own certificate, they cannot change the group list. We check that the groups are the same and if they are we update the certificate, but if the certificate is the same as the existing certificate there is no reason to perform the update.

Reply:
please could you update the comment with some of this context, ta